dashboard/app/helpers/users_helper.rb

```diff
@@ -291,52 +291,49 @@ def script_progress_for_users(users, unit)
     script.script_levels.each do |sl|
       sl.level_ids.each do |level_id|
         level = Level.cache_find(level_id)
-        users.each do |user|
-          level_for_progress = level.get_level_for_progress(user, script)
-
-          level_progress = get_level_progress(
-            user_id: user.id,
-            user_level: user_levels_by_level[user.id].try(:[], level_for_progress.id),
-            feedback_review_state:
-              teacher_feedback_by_level[user.id].try(:[], level_for_progress.id)&.review_state,
-            script_level: sl,
-            paired_user_levels: paired_user_levels[user.id],
-            include_timestamp: include_timestamp
-          )
-
-          if level.is_a?(BubbleChoice) # we have a parent level
-            bubble_choice_progress = get_bubble_choice_progress(
-              level, user, user_levels_by_level[user.id], teacher_feedback_by_level[user.id], sl, paired_user_levels[user.id], include_timestamp
+        if level.is_a?(BubbleChoice) # we have a parent level
+          # Load this once outside the users loop so the sublevels aren't
+          # queried anew for each user.
+          sublevels = level.sublevels
+          users.each do |user|
+            progress[user.id].merge!(
+              get_bubble_choice_progress(
+                level: level,
+                sublevels: sublevels,
+                user: user,
+                user_levels_by_level: user_levels_by_level[user.id],
+                teacher_feedback_by_level: teacher_feedback_by_level[user.id],
+                script_level: sl,
+                paired_user_levels: paired_user_levels[user.id],
+                include_timestamp: include_timestamp
+              )
             )
-            if bubble_choice_progress.present?
-              progress[user.id].merge!(bubble_choice_progress.compact)
-
-              sum_time_spent = bubble_choice_progress.values.reduce(0) do |sum, sublevel_progress|
-                sublevel_progress[:time_spent] ? sum + sublevel_progress[:time_spent] : sum
-              end
-
-              # The existence of level_progress needs to be checked due to a race condition where the user makes sublevel
```

Comment:
Is this race condition resolved by the refactor? Should we keep the comment if not?

Reply:
I believe it is solved, yes. Before, the progress for the parent level would be generated by querying `get_sublevel_for_progress` to figure out which sublevel to use for the summary, which happened after the queries to pull in all the progress data. Therefore it was possible that a student who had made progress on a sublevel in between those two queries would have the "right" sublevel be one that had no progress recorded in the original bulk query. Now we're using the bulk query data for both, so no race condition.